src/gui/window.rs

```diff
@@ -51,101 +56,386 @@ pub struct GuiWasabiWindow {
+                                let position = Duration::from_secs_f64(progress * length);
+                                self.midi_file.as_mut().unwrap().timer_mut().seek(position);
+                            }
+                        }
                     } else {
-                        self.midi_file.timer_mut().seek(Duration::from_secs(0));
+                        let mut progress = 0.0;
+                        let slider = egui::Slider::new(&mut progress, 0.0..=1.0).show_value(false);
+                        ui.spacing_mut().slider_width = window_size[0] - 20.0;
+                        ui.add(slider);
                     }
-                }
-            });
+                });
+            panel_height + 20.0
+        } else {
+            0.0
+        };
 
         // Calculate available space left for keyboard and notes
         // We must render notes before keyboard because the notes
         // renderer tells us the key colors
         let available = ctx.available_rect();
         let height = available.height();
-        let keyboard_height = 60.0 / 720.0 * available.width() as f32;
+        let visible_keys = std::ops::Range {
+            start: perm_settings.first_key,
+            end: perm_settings.last_key + 1,
+        }
+        .len();
+        let keyboard_height = 11.6 / visible_keys as f32 * available.width() as f32;
         let notes_height = height - keyboard_height;
 
-        let key_view = self.keyboard_layout.get_view_for_keys(0, 127);
+        let key_view = self
+            .keyboard_layout
+            .get_view_for_keys(perm_settings.first_key, perm_settings.last_key);
 
-        let no_frame = Frame::default().inner_margin(Margin::same(0.0));
+        let no_frame = Frame::default()
+            .inner_margin(Margin::same(0.0))
+            .fill(perm_settings.bg_color);
+
+        let stats_frame = Frame::default()
+            .inner_margin(egui::style::Margin::same(10.0))
+            .fill(egui::Color32::from_rgba_unmultiplied(0, 0, 0, 170))
+            .rounding(egui::Rounding::same(5.0));
 
         let mut render_result_data = None;
 
-        // Render the notes
-        egui::TopBottomPanel::top("Note panel")
-            .height_range(notes_height..=notes_height)
-            .frame(no_frame)
-            .show(&ctx, |ui| {
-                let result =
-                    self.render_scene
-                        .draw(state, ui, &key_view, &mut self.midi_file, 0.25);
-                render_result_data = Some(result);
-            });
-
-        let render_result_data = render_result_data.unwrap();
-
-        // Render the keyboard
-        egui::TopBottomPanel::top("Keyboard panel")
-            .height_range(keyboard_height..=keyboard_height)
-            .frame(no_frame)
-            .show(&ctx, |ui| {
-                self.keyboard
-                    .draw(ui, &key_view, &render_result_data.key_colors);
-            });
+        if self.midi_file.is_some() {
```

Comment:
This way of duplicating huge chunks of code between an if/else isn't really that clean and would be hard to maintain in the future.

Ideally, most of the logic should be shared and only parts of it should be behind if statements.

For if statements, it's good to use `if let Some(midi_file) = self.midi_file.as_mut() {}` instead of unwrapping.

Reply:
Yes true, I'll work a bit on duplicates.